lib/semantics/resolve-names.cc

```diff
@@ -41,27 +41,35 @@ static GenericSpec MapGenericSpec(const parser::GenericSpec &);
 // representing the implicit type; nullptr if none.
 class ImplicitRules {
 public:
-  ImplicitRules(MessageHandler &messages);
-  bool isImplicitNoneType() const { return isImplicitNoneType_; }
-  bool isImplicitNoneExternal() const { return isImplicitNoneExternal_; }
+  ImplicitRules(MessageHandler &messages)
+    : messages_{messages}, inheritFromParent_{false} {}
+  ImplicitRules(std::unique_ptr<ImplicitRules> &&parent)
+    : messages_{parent->messages_}, inheritFromParent_{true} {
+    parent_.swap(parent);
+  }
+  std::unique_ptr<ImplicitRules> &&parent() { return std::move(parent_); }
+  bool isImplicitNoneType() const;
+  bool isImplicitNoneExternal() const;
   void set_isImplicitNoneType(bool x) { isImplicitNoneType_ = x; }
   void set_isImplicitNoneExternal(bool x) { isImplicitNoneExternal_ = x; }
+  void set_inheritFromParent(bool x) { inheritFromParent_ = x; }
   // Get the implicit type for identifiers starting with ch. May be null.
   std::optional<const DeclTypeSpec> GetType(char ch) const;
   // Record the implicit type for this range of characters.
   void SetType(const DeclTypeSpec &type, parser::Location lo, parser::Location,
       bool isDefault = false);
-  // Apply the default implicit rules (if no IMPLICIT NONE).
-  void AddDefaultRules();
 
 private:
   static char Incr(char ch);
 
+  std::unique_ptr<ImplicitRules> parent_;
   MessageHandler &messages_;
-  bool isImplicitNoneType_{false};
-  bool isImplicitNoneExternal_{false};
+  std::optional<bool> isImplicitNoneType_;
```

Comment:
I think you need a comment here or near here that explains these booleans and what conditions affect inheritFromParent.

Reply:
OK, done